### PR TITLE
Add only modified subcommand for compiletest

### PR DESCRIPTION
**Pull request by chenyukang**
r? @jyn514  

From [discussion](https://rust-lang.zulipchat.com/#narrow/stream/326414-t-infra.2Fbootstrap/topic/Add.20subcommand.20to.20retest.20all.20tests.20with.20different.20results)

### CODE REVIEW

File: src/bootstrap/flags.rs [hunk 1]
```diff
@@ -301,6 +302,7 @@ To learn more about a subcommand, run `./x.py <subcommand> -h`",
                 opts.optflag("", "doc", "only run doc tests");
                 opts.optflag("", "bless", "update all stderr/stdout files of failing ui tests");
                 opts.optflag("", "force-rerun", "rerun tests even if the inputs are unchanged");
+                opts.optflag("", "only-modified", "only run tests that result has been changed");
```
Comment (chenyukang) [inline src/bootstrap/flags.rs:305]: If we don't add an sub option in top level, where is more proper, I think this `on-modified` option will only used in `compiletest`.

File: src/tools/compiletest/src/main.rs [hunk 2]
```diff
@@ -591,6 +595,24 @@ fn collect_tests_from_dir(
     let build_dir = output_relative_path(config, relative_dir_path);
     fs::create_dir_all(&build_dir).unwrap();
 
+    if config.only_modified {
+        let Ok(Some(files)) = get_git_modified_files(Some(dir), &vec!["rs", "stderr", "fixed"]) else { return Ok(()); };
+        // Add new test cases to the list, it will be convenient in daily development.
+        let Ok(Some(untracked_files)) = get_git_untracked_files(None) else { return Ok(()); };
```
Comment (chenyukang) [inline src/tools/compiletest/src/main.rs:601]: New added testcase will also run frequently, so I added it here.

**Comment by jyn514** [comment]
I don't have time for reviews right now, sorry.

r? bootstrap

**Comment by jyn514** [comment]
cc @albertlarsan68

File: src/tools/compiletest/src/main.rs [hunk 3]
```diff
@@ -591,6 +595,24 @@ fn collect_tests_from_dir(
...
+        let mut all_paths = files;
+        all_paths.extend(untracked_files);
+        let mut full_paths: Vec<PathBuf> =
+            all_paths.iter().map(|f| fs::canonicalize(&f).unwrap().with_extension("rs")).collect();
+        full_paths.dedup();
```
Comment (albertlarsan68) [inline src/tools/compiletest/src/main.rs:606]: For this `dedup()` to be fully effective, the values need to be sorted. Are they because of the guaranteed output of the git command, or is it only an optimisation that does not affect correctness?

Comment (albertlarsan68) [inline src/tools/compiletest/src/main.rs:606]:
```suggestion
        let all_paths = {
            let mut all_paths = files;
            all_paths.extend(untracked_files);
            all_paths
        };
        let full_paths = {
            let mut full_paths: Vec<PathBuf> =
                all_paths.into_iter().map(|f| fs::canonicalize(&f).unwrap().with_extension("rs")).collect();
            full_paths.dedup();
            full_paths
        };
```

This is a very OCD nit, but the less mutability there is, the better is for me (absolutely not a blocker)

**Review by albertlarsan68 — COMMENTED** [review]
Seems good, just a nit.

File: src/tools/compiletest/src/main.rs (hunk 3 above)
Comment (chenyukang) [reply src/tools/compiletest/src/main.rs:606]: Use this to concate two `vec` seems better?

```rust
let all_paths = [&files[..], &untracked_files[..]].concat();
```

Comment (chenyukang) [reply src/tools/compiletest/src/main.rs:606]: emm, I just don't realize the pre-condition of `dedup`, I think we'd better sort it.

Comment (albertlarsan68) [reply src/tools/compiletest/src/main.rs:606]: If it works, then yes.

Comment (albertlarsan68) [reply src/tools/compiletest/src/main.rs:606]: The doc is at https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.dedup

**Comment by chenyukang** [comment]
@rustbot ready

comments feedback and some code refactor.

File: src/bootstrap/format.rs [hunk 4]
```diff
@@ -169,7 +157,7 @@ pub fn format(build: &Builder<'_>, check: bool, paths: &[PathBuf]) {
                 ignore_fmt.add(&format!("!/{}", untracked_path)).expect(&untracked_path);
             }
             if !check && paths.is_empty() {
-                match get_modified_rs_files(build) {
+                match get_modified_rs_files(&build) {
```
Comment (albertlarsan68) [inline src/bootstrap/format.rs:160]: Why this change?
The build is already a borrow, thus no need to borrow the borrow.

File: src/tools/compiletest/src/main.rs [hunk 5]
```diff
@@ -597,7 +631,7 @@ fn collect_tests_from_dir(
         let file = file?;
         let file_path = file.path();
         let file_name = file.file_name();
-        if is_test(&file_name) {
+        if is_test(&file_name) && (!config.only_modified || only_modified.contains(&file_path)) {
```
Comment (albertlarsan68) [inline src/tools/compiletest/src/main.rs:634]: This means that if the `modified_test` functions errors (which is signalled by returning an empty `Vec`), then no tests are run, or am I wrong?
The old way of using `Option` was good.

Comment (chenyukang) [reply src/tools/compiletest/src/main.rs:634]: yes, the old way has a bug for the `relative_dir`.
Old way also won't run any test if it specified with `only-modified` and `git` run with error.

File: src/bootstrap/format.rs (hunk 4 above)
Comment (chenyukang) [reply src/bootstrap/format.rs]: rollback.

File: src/tools/compiletest/src/main.rs (hunk 5 above)
Comment (albertlarsan68) [reply src/tools/compiletest/src/main.rs]: Then now what is the difference between "No tests modified/added" and "An error occured"?

Comment (chenyukang) [reply src/tools/compiletest/src/main.rs]: Just added error message for error occured.

**Comment by albertlarsan68** [comment]
Thanks!
@bors r+
